qualtran/surface_code/fifteen_to_one.py

```diff
@@ -0,0 +1,380 @@
+    adds X and Z errors (storage_errors).
+    The probabilities of those channels are the same as those in the supplementary
+    material of https://arxiv.org/abs/1905.06903.
+
+    Args:
+        phys_err: physical error rate.
+        d_X: Side length of the surface code along which X measurements happen.
+        d_Z: Side length of the surface code along which Z measurements happen.
+        d_m: Number of code cycles used in lattice surgery.
+        qec: Quantum error correction scheme being used.
+
+
+    Returns:
+        The factory as a cirq circuit.
+    """
+    qs = cirq.LineQubit.range(5)
+    px = qec.logical_error_rate(d_X, phys_err)
+    pz = qec.logical_error_rate(d_Z, phys_err)
+    pm = qec.logical_error_rate(d_m, phys_err)
+
+    factory = cirq.Circuit.from_moments(
+        cirq.H.on_each(qs),
+        PauliRotationChannel(
+            'IZIII',
+            phys_err / 3 + 0.5 * (d_m / d_Z) * pz * d_m,
+            phys_err / 3 + 0.5 * d_Z * pm,
+            phys_err / 3,
+        )(
+            *qs
+        ),  # 1
+        PauliRotationChannel(
+            'IIZII',
+            phys_err / 3 + 0.5 * (d_m / d_Z) * pz * d_m,
+            phys_err / 3 + 0.5 * d_Z * pm,
+            phys_err / 3,
+        )(
+            *qs
+        ),  # 2
+        PauliRotationChannel(
+            'IIIZI',
+            phys_err / 3 + 0.5 * (d_m / d_Z) * pz * d_m,
+            phys_err / 3 + 0.5 * d_Z * pm,
+            phys_err / 3,
+        )(
+            *qs
+        ),  # 3
+        PauliRotationChannel(
+            'IZZZI',
+            phys_err / 3 + 0.5 * pm * d_m,
+            phys_err / 3 + 0.5 * pm * d_m + 0.5 * (3 * d_Z) * d_X / d_m * pm,
+            phys_err / 3,
+        )(
+            *qs
+        ),  # 5
+        *storage_error(
+            'X',
+            [
+                0,
+                0.5 * (d_Z / d_X) * px * d_m,
+                0.5 * (d_Z / d_X) * px * d_m,
+                0.5 * (d_Z / d_X) * px * d_m,
+                0,
+            ],
+            qs,
+        ),
+        *storage_error(
+            'Z',
+            [
+                0,
+                0.5 * (d_X / d_Z) * pz * d_m,
+                0.5 * (d_X / d_Z) * pz * d_m,
+                0.5 * (d_X / d_Z) * pz * d_m,
+                0,
+            ],
+            qs,
+        ),
+        PauliRotationChannel(
+            'ZZZII',
+            phys_err / 3 + 0.5 * pm * d_m,
+            phys_err / 3 + 0.5 * pm * d_m + 0.5 * (d_X + 2 * d_Z) * d_X / d_m * pm,
+            phys_err / 3,
+        )(
+            *qs
+        ),  # 6
+        PauliRotationChannel(
+            'ZZIZI',
+            phys_err / 3 + 0.5 * pm * d_m,
+            phys_err / 3 + 0.5 * pm * d_m + 0.5 * (d_X + 3 * d_Z) * d_X / d_m * pm,
+            phys_err / 3,
+        )(
+            *qs
+        ),  # 7
+        *storage_error(
+            'Z', [0.5 * ((d_X + 2 * d_Z) + (d_X + 3 * d_Z)) / d_X * px * d_m, 0, 0, 0, 0], qs
+        ),
+        *storage_error(
```

Comment:
why do the storage errors only come after certain groups of the rotations? 

Can you factor out the storage_error operations? Looks like a lot of code duplication here. e.g.

```python

x_storage_err = get_storage_error('X')
z_storage_err = get_storage_error('Z')
cirq.Circuit([
  Pauli...
  Pauli...
  x_storage_err,
  z_storage_err,
  PauliRot
  ...
```

Reply:
they don't explain why they grouped them in this way in the paper.  the storage errors have different sets of probabilities. so they can't be extracted